File: junior-backend/app/analytics.py
"""PostHog event emitter for the backend.

Privacy posture (reviewer spec):
  - Identify with internal IDs only: user.id (our UUID), clerk_id,
    whop_user_id, affiliate_id.
  - NEVER send: raw email, JWTs, access tokens, license tokens, Whop
    access tokens, local paths, video filenames, transcripts.
  - PostHog is observability only — the DB is the source of truth for
    affiliate attribution and payouts.

We construct the client lazily so missing config never breaks an import,
and we use the project key (POSTHOG_KEY) — there's no need for a
personal API key for event capture.
"""

from __future__ import annotations

import logging
from typing import Any, Literal

from app.config import get_settings

log = logging.getLogger("junior.analytics")

_client = None
_inited = False

# Canonical affiliate-flywheel funnel chain (CLAUDE_POSTHOG_AFFILIATE_FUNNEL_
# ARCHITECTURE.md). Listed here as the single source of truth for event names;
# the comment marks where each one is emitted. Phase 1 wires only the events
# that naturally happen today — the rest are reserved names for later phases.
#
#   reward_clip_viewed              → backend sync/manual import   (Phase 3)
#   affiliate_link_clicked          → marketing/account frontend   (Phase 1)
#   starter_pass_started            → backend ledger redeem        (Phase 3)
#   desktop_activated               → backend /desktop/connect     (Phase 1 ✓ live)
#   first_bounty_workspace_created  → desktop sidecar              (Phase 1)
#   bounty_clip_exported            → desktop sidecar              (Phase 1)
#   starter_pass_exhausted          → backend ledger/cron          (Phase 3)
#   subscription_activated          → backend clerk webhook        (Phase 1 ✓ live)
#   subscription_still_active_day_30 → backend scheduled job       (Phase 2)
BackendEvent = Literal[
    # Webhook lifecycle
    "signup_completed",                 # user.created handled
    "affiliate_attribution_locked",     # affiliate_id baked at signup
    "whop_membership_valid",            # paid subscription went live
    "pending_whop_membership_stashed",  # buyer paid before signup; entitlement parked
    "whop_trial_started",               # membership valid → trialing/starter state
    "whop_payment_succeeded",           # payment promoted user to active paid
    "subscription_activated",           # Clerk subscription_active
    "subscription_canceled",
    "subscription_still_active_day_30", # retention check (Phase 2 scheduled job)
    # Desktop activation
    "desktop_activated",                # /desktop/connect first license issued
    # Affiliate starter-pass flywheel (names reserved; emit homes per Phase)
    "reward_clip_viewed",               # Whop reward view sync (Phase 3)
    "starter_pass_started",             # ledger redeem (Phase 3)
    "starter_pass_exhausted",           # credits/expiry (Phase 3)
    # Self-serve claim (bought on Whop with a different email than signup)
    "whop_claim_email_sent",            # claim link emailed to the Whop purchase email
    "whop_claim_succeeded",             # claim token redeemed → pending applied
    "whop_claim_failed",                # invalid/expired/used token or mismatch
]


def _ensure_client() -> Any:
    global _client, _inited
    if _inited:
        return _client
    s = get_settings()
    if not s.posthog_key:
        _inited = True
        return None
    try:
        import posthog as ph
        ph.api_key = s.posthog_key
        ph.host = s.posthog_host
        _client = ph
    except Exception as e:  # noqa: BLE001
        log.warning("[analytics] PostHog client init failed: %s", e)
        _client = None
    _inited = True
    return _client


# Property keys we always strip before sending. Defence-in-depth for the
# call sites that should already follow the rules.
_FORBIDDEN_KEYS = frozenset(
    {
        "email", "user_email", "primary_email",
        "token", "access_token", "id_token", "jwt", "license_jwt",
        "api_key", "secret", "password",
        "path", "filename", "source_path", "source_filename",
        "transcript",
    }
)


def _sanitize(props: dict[str, Any] | None) -> dict[str, Any]:
    if not props:
        return {}
    out: dict[str, Any] = {}
    for k, v in props.items():
        if k.lower() in _FORBIDDEN_KEYS:
            continue
        out[k] = v
    return out


def capture(
    *,
    user_id: str,
    event: BackendEvent,
    properties: dict[str, Any] | None = None,
) -> None:
    """Emit a backend event. `user_id` is whatever string the frontend used
    to identify the user — for us that's the Clerk user id (`user.clerk_id`).
    Keeps frontend and backend events on the same person without mixing in
    our internal UUID."""
    client = _ensure_client()
    if not client:
        return
    try:
        client.capture(
            distinct_id=user_id,
            event=event,
            properties=_sanitize(properties),
        )
    except Exception as e:  # noqa: BLE001
        log.warning("[analytics] capture failed event=%s err=%s", event, e)


def identify(
    *,
    user_id: str,
    clerk_id: str | None = None,
    affiliate_id: str | None = None,
    whop_user_id: str | None = None,
    tier: str | None = None,
) -> None:
    """Set person properties so funnel dashboards can group by tier /
    affiliate without us shipping the email."""
    client = _ensure_client()
    if not client:
        return
    try:
        client.set(
            distinct_id=user_id,
            properties={
                k: v
                for k, v in {
                    "clerk_id": clerk_id,
                    "affiliate_id": affiliate_id,
                    "whop_user_id": whop_user_id,
                    "tier": tier,
                }.items()
                if v is not None
            },
        )
    except Exception as e:  # noqa: BLE001
        log.warning("[analytics] identify failed user=%s err=%s", user_id, e)

File: account-app/src/lib/analytics.ts
// PostHog — observability only. Backend DB is the source of truth for
// attribution; these events tell us where the funnel breaks, not who gets paid.
//
// Privacy rules (reviewer spec):
//   - Identify with internal IDs only: clerk_id / affiliate_id / whop_user_id
//   - NEVER send: raw email, access tokens, license JWTs, local paths,
//     video filenames, transcripts, Whop access tokens
//
// Project key (NEXT_PUBLIC_POSTHOG_KEY) is public-safe by design. The
// secret-style "personal API key" PostHog also offers is NOT used here — we
// only need event capture.

"use client";

import posthog, { type PostHog } from "posthog-js";

const KEY = process.env.NEXT_PUBLIC_POSTHOG_KEY;
const HOST = process.env.NEXT_PUBLIC_POSTHOG_HOST ?? "https://us.i.posthog.com";

let inited = false;

export function initAnalytics(): PostHog | null {
  if (typeof window === "undefined") return null;
  if (!KEY) return null;
  if (inited) return posthog;
  posthog.init(KEY, {
    api_host: HOST,
    capture_pageview: false, // We send our own page events with stable names
    capture_pageleave: true,
    persistence: "localStorage+cookie",
    // Default would send `$current_url` with raw URLs. Strip query strings on
    // marketing pages so an affiliate code (?a=…) doesn't ride along as PII.
    sanitize_properties: (props) => {
      if (props && typeof props.$current_url === "string") {
        try {
          const u = new URL(props.$current_url);
          // Keep host + path; drop search + hash (search may contain ?a=, ?ref=, etc.)
          props.$current_url = `${u.origin}${u.pathname}`;
        } catch {
          /* leave as-is */
        }
      }
      return props;
    },
  });
  inited = true;
  return posthog;
}

export type IdentifyContext = {
  clerk_id: string;
  affiliate_id?: string | null;
  tier?: string;
  whop_user_id?: string | null;
  // No email. No name. No platform handles.
};

export function identifyUser(ctx: IdentifyContext): void {
  if (!inited) return;
  posthog.identify(ctx.clerk_id, {
    // Only the IDs and tier. Email is intentionally absent.
    affiliate_id: ctx.affiliate_id ?? undefined,
    tier: ctx.tier ?? undefined,
    whop_user_id: ctx.whop_user_id ?? undefined,
  });
}

export function track(event: AnalyticsEvent, properties?: Record<string, unknown>): void {
  if (!inited) return;
  posthog.capture(event, sanitize(properties));
}

export function resetAnalytics(): void {
  if (!inited) return;
  posthog.reset();
}

// --- safe property filter -----------------------------------------------
// Defence-in-depth: even though every call site is meant to send only
// allowed properties, we strip anything that looks like PII or a secret
// before it leaves the browser.

const FORBIDDEN_KEYS = new Set([
  "email", "user_email", "primary_email",
  "token", "access_token", "id_token", "jwt", "license_jwt",
  "api_key", "secret", "password",
  "path", "filename", "transcript", "source_filename", "source_path",
]);

function sanitize(props?: Record<string, unknown>): Record<string, unknown> | undefined {
  if (!props) return undefined;
  const out: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(props)) {
    if (FORBIDDEN_KEYS.has(k.toLowerCase())) continue;
    out[k] = v;
  }
  return out;
}

// --- the closed event vocabulary ---------------------------------------
// One enum for the whole app — adding a new event = adding to this union.
// PostHog dashboards reference these names verbatim.

export type AnalyticsEvent =
  // Affiliate funnel (marketing surface in this app for now)
  | "affiliate_landing_viewed"
  | "affiliate_ref_captured"
  // Top-of-funnel: a referred visitor reached the app (ref cookie/param seen),
  // fired once per session before they reach the signup form. Distinct from
  // affiliate_ref_captured (which fires on the signup surface itself).
  | "affiliate_link_clicked"
  // Signup → activated
  | "signup_started"
  | "signup_completed"
  | "dashboard_viewed"
  | "desktop_download_clicked"
  // Billing
  | "checkout_started"
  | "upgrade_viewed"
  // Connections
  | "whop_connect_clicked"
  | "connection_added";
